test/e2e/cpi_vm_test.go

```diff
@@ -0,0 +1,177 @@
+func getInternalIPFromNode(node *corev1.Node) (string, error) {
+	addresses := node.Status.Addresses
+	for _, address := range addresses {
+		if address.Type == corev1.NodeInternalIP {
+			return address.String(), nil
+		}
+	}
+	return "", errors.New("internal IP not found")
+}
+
+// getProviderIPFromNode returns the provider ID of node
+func getProviderIPFromNode(node *corev1.Node) string {
+	return node.Spec.ProviderID
+}
+
+// IsNodeReady returns whether the not is the given node ready
+func IsNodeReady(node *corev1.Node) bool {
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == corev1.NodeReady {
+			return condition.Status == corev1.ConditionTrue
+		}
+	}
+	return false
+}
+
+var _ = Describe("Recreating VMs", func() {
+	It("should result in new nodes", func() {
+		// TODO(fhan)
+	})
+})
+
+/*
+	Restart a worker node, then assert that the external, internal IP and
+	the provider ID for the node should not change.
+*/
+var _ = Describe("Restarting VMs", func() {
+
+	var workerNode *corev1.Node
+	var workerVM *object.VirtualMachine
+
+	BeforeEach(func() {
+		By("Get the name of worker node", func() {
+			workerNode, err = getWorkerNode()
+			Expect(err).ToNot(HaveOccurred())
+
+			klog.Infof("The worker node for testing is %s\n", workerNode.Name)
+		})
+
+		By("Get corresponding VM object for node", func() {
+			workerVMs, err := vsphere.Finder.VirtualMachineList(ctx, workerNode.Name)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(workerVMs).To(HaveLen(1))
+			workerVM = workerVMs[0]
+		})
+	})
+
+	It("should pertain the original node when VM restarts", func() {
+		By("Shutdown VM "+workerVM.Name(), func() {
+			task, err := workerVM.PowerOff(ctx)
+			Expect(err).ToNot(HaveOccurred(), "cannot power off vm")
+
+			err = task.Wait(ctx)
+			Expect(err).ToNot(HaveOccurred(), "cannot wait for vm to power off")
+		})
+
+		externalIP, err := getExternalIPFromNode(workerNode)
+		Expect(err).ToNot(HaveOccurred())
+
+		internalIP, err := getInternalIPFromNode(workerNode)
+		Expect(err).ToNot(HaveOccurred())
+
+		providerID := getProviderIPFromNode(workerNode)
+
+		Expect(IsNodeReady(workerNode)).To(BeTrue())
+
+		By("Wait for VM " + workerVM.Name() + " to go down")
+		Eventually(func() error {
+			workerNode, err = getWorkerNode()
+			if err != nil {
+				return err
+			}
+			if IsNodeReady(workerNode) {
+				return errors.New("worker node hasn't gone offline yet")
+			}
+			return nil
+		}, 5*time.Minute, 2*time.Second).Should(BeNil())
+
+		By("Power on VM "+workerVM.Name(), func() {
+			task, err := workerVM.PowerOn(ctx)
+			Expect(err).ToNot(HaveOccurred(), "cannot power on vm")
+
+			err = task.Wait(ctx)
+			Expect(err).ToNot(HaveOccurred(), "cannot wait for vm to power on")
+		})
+
+		By("Wait for VM " + workerVM.Name() + " to go up again")
+		Eventually(func() error {
+			workerNode, err = getWorkerNode()
+			if err != nil {
+				return err
+			}
+			if !IsNodeReady(workerNode) {
```

Comment:
Here it make sense to check node's ready, because VM is up doesn't mean node will be ready

Reply:
added VM power state check as well